From de0cd1a68bb0abb0ef5a2838b8cad9c926aac5b0 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Mon, 26 Oct 2015 09:41:41 +0900
Subject: [PATCH 1897/2066] media: i2c: adv7482: Add input resolution
 detection dynamically

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
[zou:Original patch taken from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git
 v4.9/rcar-3.5.1]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 drivers/media/i2c/adv7482.c |   63 ++++++++++++++++++++++++++++++------------
 1 files changed, 45 insertions(+), 18 deletions(-)

diff --git a/drivers/media/i2c/adv7482.c b/drivers/media/i2c/adv7482.c
index ff6885d..ec7d05b 100644
--- a/drivers/media/i2c/adv7482.c
+++ b/drivers/media/i2c/adv7482.c
@@ -908,9 +908,11 @@ static int adv7482_get_vid_info(struct v4l2_subdev *sd, u8 *progressive,
 static int adv7482_set_vid_info(struct v4l2_subdev *sd)
 {
 	struct adv7482_state *state = to_state(sd);
+	struct i2c_client *client = v4l2_get_subdevdata(sd);
 
 	u8 progressive;
 	u8 signal;
+	u8 val = ADV7482_IO_CP_VID_STD_480P;
 	u32 width;
 	u32 height;
 	int ret;
@@ -921,33 +923,58 @@ static int adv7482_set_vid_info(struct v4l2_subdev *sd)
 		width		= ADV7482_MAX_WIDTH;
 		height		= ADV7482_MAX_HEIGHT;
 		progressive	= 1;
-	}
-	/*
-	   FIXME : when hw detect active resolution, it's set it.
-	*/
-	else
-		if ((width == 720) &&
-			(height == 480) && (progressive == 1))
+	} else {
+		if ((width == 640) &&
+			(height == 480) && (progressive == 1)) {
+			val = ADV7482_IO_CP_VID_STD_VGA60;
+			dev_info(state->dev,
+				 "Changed active resolution to 640x480p\n");
+		} else if ((width == 720) &&
+			(height == 480) && (progressive == 1)) {
+			val = ADV7482_IO_CP_VID_STD_480P;
 			dev_info(state->dev,
 				 "Changed active resolution to 720x480p\n");
-		else if ((width == 720) &&
-			(height == 480) && (progressive == 0))
+		} else if ((width == 720) &&
+			(height == 480) && (progressive == 0)) {
+			val = ADV7482_IO_CP_VID_STD_480I;
 			dev_info(state->dev,
 				 "Changed active resolution to 720x480i\n");
-		else if ((width == 1920) &&
-			(height == 1080) && (progressive == 1))
+		} else if ((width == 720) &&
+			(height == 576) && (progressive == 1)) {
+			val = ADV7482_IO_CP_VID_STD_576P;
 			dev_info(state->dev,
-				 "Changed active resolution to 1920x1080p\n");
-		else if ((width == 1920) &&
-			(height == 1080) && (progressive == 0))
+				 "Changed active resolution to 720x576p\n");
+		} else if ((width == 720) &&
+			(height == 576) && (progressive == 0)) {
+			val = ADV7482_IO_CP_VID_STD_576I;
 			dev_info(state->dev,
-				 "Changed active resolution to 1920x1080i\n");
-		else if ((width == 1280) &&
-			(height == 720) && (progressive == 1))
+				 "Changed active resolution to 720x576i\n");
+		} else if ((width == 1280) &&
+			(height == 720) && (progressive == 1)) {
+			val = ADV7482_IO_CP_VID_STD_720P;
 			dev_info(state->dev,
 				 "Changed active resolution to 1280x720p\n");
+		} else if ((width == 1920) &&
+			(height == 1080) && (progressive == 1)) {
+			val = ADV7482_IO_CP_VID_STD_1080P;
+			dev_info(state->dev,
+				 "Changed active resolution to 1920x1080p\n");
+		} else if ((width == 1920) &&
+			(height == 1080) && (progressive == 0)) {
+			val = ADV7482_IO_CP_VID_STD_1080I;
+			dev_info(state->dev,
+				 "Changed active resolution to 1920x1080i\n");
+		} else {
+			dev_err(state->dev,
+				 "Not support resolution %dx%d%c\n",
+				  width, height, (progressive) ? 'p' : 'i');
+			return -EINVAL;
+		}
+	}
+	ret = adv7482_write_register(client, ADV7482_I2C_IO,
+				ADV7482_IO_CP_VID_STD_REG, val);
 
-	return 0;
+	return ret;
 }
 
 /*****************************************************************************/
-- 
1.7.5.4

